dev/bots/analyze_snippet_code.dart

```diff
@@ -431,346 +725,260 @@ class _SnippetChecker {
-class _Section {
-  const _Section(this.code, {this.dartVersionOverride});
-  factory _Section.combine(List<_Section> sections) {
-    final List<_Line> code = sections
-        .expand((_Section section) => section.code)
-        .toList();
-    return _Section(code);
-  }
-  factory _Section.fromStrings(_Line firstLine, List<String> code) {
-    final List<_Line> codeLines = <_Line>[];
-    for (int i = 0; i < code.length; ++i) {
-      codeLines.add(
-        _Line(
-          code: code[i],
-          filename: firstLine.filename,
-          line: firstLine.line + i,
-          indent: firstLine.indent,
-        ),
-      );
+/// A class to represent a section of snippet code, marked by "```dart ... ```", that ends up
+/// in a file we then analyze (each snippet is in its own file).
+class _SnippetFile {
+  const _SnippetFile(this.code, this.generatorComment, this.filename, this.indexLine);
+
+  factory _SnippetFile.fromLines(
+    List<_Line> code,
+    List<_Line> headers,
+    List<_Line> preamble,
+    String generatorComment,
+    String filename,
+  ) {
+    while (code.isNotEmpty && code.last.code.isEmpty) {
+      code.removeLast();
     }
-    return _Section(codeLines);
+    assert(code.isNotEmpty);
+    final _Line firstLine = code.firstWhere((_Line line) => !line.generated);
+    return _SnippetFile(
+      <_Line>[
+        ...headers,
+        const _Line.generated(), // blank line
+        if (preamble.isNotEmpty)
+          ...preamble,
+        if (preamble.isNotEmpty)
+          const _Line.generated(), // blank line
+        _Line.generated(code: '// From: $filename:${firstLine.line}'),
+        ...code,
+      ],
+      generatorComment,
+      filename,
+      firstLine.line,
+    );
   }
-  factory _Section.surround(_Line firstLine, String prefix, List<String> code, String postfix) {
-    assert(prefix != null);
-    assert(postfix != null);
+
+  factory _SnippetFile.fromStrings(
+    _Line firstLine,
+    List<String> code,
+    List<_Line> headers,
+    List<_Line> preamble,
+    String generatorComment,
+    String filename, {
+    String? prefix, String? postfix,
+  }) {
     final List<_Line> codeLines = <_Line>[];
+    if (prefix != null) {
+      codeLines.add(_Line.generated(code: prefix));
+    }
     for (int i = 0; i < code.length; ++i) {
       codeLines.add(
         _Line(
           code: code[i],
-          filename: firstLine.filename,
           line: firstLine.line + i,
           indent: firstLine.indent,
         ),
       );
     }
-    return _Section(<_Line>[
-      _Line.generated(code: prefix, filename: firstLine.filename, line: 0),
-      ...codeLines,
-      _Line.generated(code: postfix, filename: firstLine.filename, line: 0),
-    ]);
+    if (postfix != null) {
+      codeLines.add(_Line.generated(code: postfix));
+    }
+    return _SnippetFile.fromLines(codeLines, headers, preamble, generatorComment, filename);
```

Comment:
We can use collection-if and collection-for:

```dart
final List<_Line> codeLines = <_Line>[
  if (prefix != null) _Line.generated(code: prefix),
  for (int i = 0; i < code.length; ++i)
    _Line(
      code: code[i],
      line: firstLine.line + i,
      indent: firstLine.indent,
    ),
  if (postfix != null) _Line.generated(code: postfix),
];
```

Reply:
done